### ISSUE TITLE
header in section with role should not have role banner

### ISSUE THREAD

**Issue by lexanth** [dequelabs/axe-core#3752]
### Product

axe-core

### Product Version

4.4.1

### Latest Version

- [X] I have tested the issue with the latest version of the product

#### Expectation
Given a dialog which is labelled by a header:
```html
<section role="dialog" aria-labelledby="dialog-header">
  <header id="dialog-header">My header</header>
  <div>Some dialog content</div>
</section>
```
I would expect the accessible name of the dialog to be identified as "My header"

#### Actual
I get a violation on "ARIA dialog and alertdialog nodes should have an accessible name (aria-dialog-name)"

On further investigation:
* It is correctly identifying the `aria-labelledby` link
* In determining the text of the `header`, it first considers its role
* The role is identified as `banner`, because https://github.com/dequelabs/axe-core/blob/522bdfb57e/lib/commons/standards/implicit-html-roles.js#L17 only links the header to the section if the section does not have a role attribute.
* Banners are not named from contents (https://github.com/dequelabs/axe-core/blob/522bdfb57e/lib/commons/aria/named-from-contents.js#L36) so the contents are ignored, so the element linked by `aria-labelledby` is identified as empty, hence the violation

#### How to Reproduce
See HTML in Expectation

#### Additional context
I haven't verified if a screen reader identifies the header as the dialog header.

I haven't seen anywhere that says that sectioning HTML elements are only sectioning when they don't have a role. https://www.w3.org/TR/2010/WD-html5-20101019/content-models.html#sectioning-content-0

All of the valid roles for a section in https://www.w3.org/TR/2010/WD-html5-20101019/content-models.html#annotations-for-assistive-technology-products-aria seem like they should still act as sectioning elements. I think invalid roles would be caught by the corresponding rule?

I can work around this by explicitly setting `role="none"` on my `header`.

**Comment by straker** [comment]
Interesting. So testing this out:

```html
<section tabindex="0" role="dialog" aria-labelledby="dialog-header">
    <div>Some dialog content</div>
  </section>

<header id="dialog-header" hidden>My header</header>
```

Each Browser / Screen Reader reports the `[role=dialog]` as "My header, dialog".

So even though [the `header` should be a banner role](https://www.w3.org/TR/html-aria/#el-header) since it's not a descendant of another sectioning element, seems screen readers are allowing named from contents on it even though it's suppose to be named from author.

**Comment by WilcoFiers** [comment]
Update: Axe-core has a bug in how we compute accessible name. Fix should be released in a few weeks in  4.8.0.

**Comment by padmavemulapati** [comment]
Verified with the latest axe-core develop branch code base ,

Previously axe-core was only considering the first part of this sentence, and skipping name from author elements while following aria-labelledby and <label> IDREFs. This was a problem on dialogs (as the issue reported), but had broader implications too. Notably, column and rowheaders in a table could not be referenced with aria-labelledby.

Code following was failed for `aria-dialogname` because of considering the part of the sentence and with the same code after fix is available it is passing 

```
<html lang="en">
  <div id="Dialog-name" role="dialog" aria-labelledby="dialog-heading">
    <header id="dialog-heading">Named from author element</header>
  </div>
```


https://github.com/dequelabs/axe-core/assets/47409579/997e2298-bba9-4d98-9728-42e65f16d390
